src/transformers/modeling_utils.py

```diff
@@ -148,6 +150,304 @@ def find_tensor_attributes(module: nn.Module) -> List[Tuple[str, Tensor]]:
+            (like `"5MB"`).
+    """
+    max_size = convert_file_size_to_int(max_size)
+
+    sharded_state_dicts = []
+    current_block = {}
+    current_block_size = 0
+    total_size = 0
+
+    for key, weight in state_dict.items():
+        weight_size = weight.numel() * dtype_size(weight.dtype)
+
+        # If this weight is going to tip up over the maximal size, we split.
+        if current_block_size + weight_size > max_size:
+            sharded_state_dicts.append(current_block)
+            current_block = {}
+            current_block_size = 0
+
+        current_block[key] = weight
+        current_block_size += weight_size
+        total_size += weight_size
+
+    # Add the last block
+    sharded_state_dicts.append(current_block)
+    return sharded_state_dicts, total_size
+
+
+def save_and_shard_checkpoint(
+    save_directory: Union[str, os.PathLike],
+    state_dict: Dict[str, torch.Tensor],
+    max_size: Union[int, str] = "5GB",
+    save_function: Callable = torch.save,
+):
+    """
+    Shards a model state dictionary and saves it in a given directory.
+
+    Arguments:
+        save_directory (`str` or `os.PathLike`):
+            Directory to which to save. Will be created if it doesn't exist.
+        state_dict (dictionary of `torch.Tensor`):
+            The state dictionary of the model to save.
+        max_shard_size (`int` or `str`, *optional*, defaults to `"10GB"`):
+            The maximum size for a checkpoint before being sharded. Checkpoints shard will then be each of size lower
+            than this size. If expressed as a string, needs to be digits followed by a unit (like `"5MB"`).
+
+            <Tip warning={true}>
+
+            If a single weight of the model is bigger than `max_shard_size`, it will be in its own checkpoint shard
+            which will be bigger than `max_shard_size`.
+
+            </Tip>
+
+        save_function (`Callable`):
+            The function to use to save the state dictionary. Useful on distributed training like TPUs when one needs
+            to replace `torch.save` by another method.
+    """
+    shards, total_size = shard_checkpoint(state_dict, max_size=max_size)
+    # One shard only means the whole model doesn't exceed the maximum size.
+    if len(shards) == 1:
+        save_file = os.path.join(save_directory, WEIGHTS_NAME)
+        logger.info(f"Model weights saved in {save_file}")
+        save_function(shards[0], save_file)
+        return
+
+    # Otherwise we build the index.
+    save_index_file = os.path.join(save_directory, WEIGHTS_INDEX_NAME)
+    weight_map = {}
+    for idx, shard in enumerate(shards):
+        shard_file = WEIGHTS_NAME.replace(".bin", f"-{idx+1:05d}-of-{len(shards):05d}.bin")
+        save_function(shard, os.path.join(save_directory, shard_file))
+        for key in shard.keys():
+            weight_map[key] = shard_file
```

Comment:
nice the keys are the actually PyTorch parameter names here no? That's cool - then the user knows which layer names are in which shard

Reply:
Yes, those are the full parameter names as in the state dict :-)